File: HW1/hate-crimes.c
#include <stdio.h>
#include <float.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

char* FILE_NAME = "./hate-crimes.csv";
int MAX_LINE_LEN = 100;

int ROW_COUNT = 51;
char* ROW_PATTERN="%s,%d,%f,%f,%f,%f,%f,%f,%f,%f,%f,%f";

/**
    Mistakes I made while writing: 

    1. -> for structs
    2. Check the max line len
    3. Make sure we have doubles instead of floats
    4. %s doesn't respect spaces, so replace spaces with underlines
    5. Fill 0 values in the data
    6. Various missing semicolons
    7. tried to separate things into files but forgot how to make it
    8. Tried to qsort but ran into junk with pointers
    9. Tokenization was a ...hassle

*/ 

struct Row {
    char *state; 
    int median_income;
    float unemployed_seasonal_percent;
    float metro_dwellers_percent;
    float high_school_grad_percent;
    float non_citizen_percent;
    float poor_white_percent;
    float gini_index;
    float non_white_percent;
    float trump_voter_percent;
    // number per 100k, use double so we can capture full decimal place
    double crimes_splc;
    double crimes_fbi;
};

void sort_names(struct Row* array[], int count) {
    bool flag;
    int i, j;
    struct Row* t;
    for (i = 1 ; i <= count - 1; i++) {
        t = array[i];

        for (j = i - 1 ; j >= 0; j--) {
            if (strcmp(array[j]->state, t->state) < 0) {
                break;  
            }

            array[j+1] = array[j];
            flag = 1;
        }

        if (flag) {
            array[j+1] = t;
        }
    }
}

void sort_income(struct Row* array[], int count) {
    bool flag;
    int i, j;
    struct Row* t;
    for (i = 1 ; i <= count - 1; i++) {
        t = array[i];

        for (j = i - 1 ; j >= 0; j--) {
            if (array[j]->median_income > t->median_income) {
                break;  
            }

            array[j+1] = array[j];
            flag = 1;
        }

        if (flag) {
            array[j+1] = t;
        }
    }
}

void read_data(struct Row* rows[]) {
    int i = 0;
    FILE *fp;

    fp = fopen(FILE_NAME, "r");
    if(fp == NULL) {
        printf("Error opening file\n");
        exit(1);
    }

    // skip through the heading line
    while (fgetc(fp) != '\n') {
        continue;
    }

    const char delimeter[2] = ",";
    char *token = malloc(sizeof(char) * MAX_LINE_LEN);
    char buffer[MAX_LINE_LEN];
    for (i = 0; i < ROW_COUNT; i++) {
        struct Row* row = rows[i];
        row->state = malloc(sizeof(char) * 20);
        fgets(buffer, MAX_LINE_LEN, fp);

        token = strtok(buffer, delimeter); 
        strcpy(row->state, token);

        token = strtok(NULL, delimeter);
        row->median_income = atoi(token); 

        token = strtok(NULL, delimeter);
        row->unemployed_seasonal_percent = atof(token);

        token = strtok(NULL, delimeter);
        row->metro_dwellers_percent = atof(token);

        token = strtok(NULL, delimeter);
        row->high_school_grad_percent = atof(token);

        token = strtok(NULL, delimeter);
        row->non_citizen_percent = atof(token);

        token = strtok(NULL, delimeter);
        row->poor_white_percent = atof(token);

        token = strtok(NULL, delimeter);
        row->gini_index = atof(token);

        token = strtok(NULL, delimeter);
        row->non_white_percent = atof(token);

        token = strtok(NULL, delimeter);
        row->trump_voter_percent = atof(token);

        token = strtok(NULL, delimeter);
        row->crimes_splc = atof(token);

        token = strtok(NULL, delimeter);
        row->crimes_fbi = atof(token);
    }
}

void print_usage(char *name) {
    printf("\nUSAGE\n---\n");
    printf("%s <compare_field>\n", name);
    printf("\t compare_field - name | median_income");
}

int main(int argc, char** argv) {
    struct Row* rows[ROW_COUNT];
    int i = 0;

    if (argc < 2) {
        printf("No argument provided for sorting.\n");
        print_usage(argv[0]);
        exit(2);
    }

    for (i = 0; i < ROW_COUNT; i++){
        rows[i] = (struct Row *) malloc(sizeof(struct Row));
    }

   read_data(rows);

    if (strcmp(argv[1], "name") == 0) {
        sort_names(rows, ROW_COUNT);
    } else if(strcmp(argv[1], "median_income") == 0) {
        sort_income(rows, ROW_COUNT);
    } else {
        printf("Unknown option for compare_field: %s\n", argv[1]);
        print_usage(argv[0]);
    }

    printf("%-20s\t INCOME\t SPLC\t FBI\n", "STATE");
    printf("---------------------------------------------\n");
    for (i = 0; i < ROW_COUNT; i++) {
        printf("%-20s\t %6d\t %.2f\t%4.2f\n", 
            rows[i]->state,
            rows[i]->median_income, 
            rows[i]->crimes_splc, 
            rows[i]->crimes_fbi
        );
    }
}

File: HW1/hate-crimes.js
/*
  Q3) The two valid compare options are name and median_income.
   
  Q4) When I re-ran the program with name as the compare option, it produced the table of states and their associated data, 
  with the states in alphabetical order.
*/

// (1) @task Import the fs library and assign it to a variable called fs
const process = require("process");
const path = require("path");
const fs = require("fs");

// (2) @task Create a variable called FILE_NAME and assign it the
// string value of hate-crimes.csv
const FILE_NAME = "./hate-crimes.csv";

const Row = function () {
  return {
    name: "",
    median_income: 0,
    unemployed_seasonal_percent: 0,
    metro_dwellers_percent: 0,
    high_school_grad_percent: 0,
    non_citizen_percent: 0,
    poor_white_percent: 0,
    gini_index: 0,
    non_white_percent: 0,
    trump_voter_percent: 0,
    crimes_splc: 0,
    crimes_fbi: 0,
  };
};

const read_data = function (rows) {
  // (3) Create a mutable variable data
  let data;
  try {
    data = fs.readFileSync(FILE_NAME, "utf-8");
  } catch (e) {
    console.log("Error opening file" + e);
    process.exit(1);
  }

  const lines = data.split("\n");
  // skip through heading line by starting at index 1
  for (let i = 1; i < lines.length; i++) {
    const row = Row();
    const rawData = lines[i].split(",");

    // ew we can do better than this
    row.name = rawData[0];
    row.median_income = parseInt(rawData[1], 10);
    row.unemployed_seasonal_percent = parseFloat(rawData[2], 10);
    row.metro_dwellers_percent = parseFloat(rawData[3], 10);
    row.high_school_grad_percent = parseFloat(rawData[4], 10);
    row.non_citizen_percent = parseFloat(rawData[5], 10);
    row.poor_white_percent = parseFloat(rawData[6], 10);
    row.gini_index = parseFloat(rawData[7], 10);
    row.non_white_percent = parseFloat(rawData[8], 10);
    row.trump_voter_percent = parseFloat(rawData[9], 10);
    row.crimes_splc = parseFloat(rawData[10], 10);
    row.crimes_fbi = parseFloat(rawData[11], 10);

    // insert another element into the array
    rows.push(row);
  }
};

const print_usage = function () {
  console.log("\nUSAGE\n---\n");
  console.log(`${path.basename(process.argv[1])} <compare_field>`);
  console.log("\t compare_field - name | median_income");
};

const main = function () {
  if (process.argv.length < 3) {
    console.log("No argument provided for sorting.");
    print_usage();
    process.exit(2);
  }

  // (4) @task Create an empty array called rows
  const rows = [];
  read_data(rows);

  const defaultRow = Row();
  const option = process.argv[2];
  if (defaultRow.hasOwnProperty(option)) {
    rows.sort(function (a, b) {
      return b[option] - a[option];
    });
  } else {
    console.log(`Unknown option for compare_field: ${option}`);
    print_usage();
    process.exit(1);
  }

  console.log(`${"STATE".padEnd(20)}\tINCOME\tSPLC\tFBI`);
  console.log("-".repeat(45));
  // (5) @task Complete the for loop using the condition that
  // the loop variable must be less than the expression rows.length
  for (let i = 0; i < rows.length; i++) {
    const row = [
      rows[i].name.padEnd(20),
      // access the median_income property of an element in the array
      rows[i].median_income,
      rows[i].crimes_splc.toFixed(2),
      // access the crimes_fbi property of an element and
      // set it to be a fixed with of two
      rows[i].crimes_fbi.toFixed(2),
    ].join("\t");

    console.log(row);
  }
};

// (6) Call the function known as main
main();
